### PR TITLE
DNS cache maintenance

### PR DESCRIPTION
**Pull request by rvansa**
This draft PR contains a backport of issue that prevents compilation on systems with newer glibc versions; ideally the backport should be integrated first and then this PR can be rebased.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Anton Kozlov](https://openjdk.org/census#akozlov) (@AntonKozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/crac pull/40/head:pull/40` \
`$ git checkout pull/40`

Update a local copy of the PR: \
`$ git checkout pull/40` \
`$ git pull https://git.openjdk.org/crac pull/40/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 40`

View PR using the GUI difftool: \
`$ git pr show -t 40`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/crac/pull/40.diff">https://git.openjdk.org/crac/pull/40.diff</a>

</details>

### CODE REVIEW

**Comment by jankratochvil** [comment]
That glibc fix I did submit but it was rejected: https://github.com/openjdk/crac/pull/37#issuecomment-1333650424

File: test/jdk/jdk/crac/ResolveInetAddress.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class ResolveInetAddress {
```
Comment (AntonKozlov) [inline test/jdk/jdk/crac/ResolveInetAddress.java:28]: As this is not a test itself, should this be moved into a subdir along with TestInetAddress.java?

File: test/jdk/jdk/crac/TestInetAddress.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    private static void ensureContainerDead() throws Exception {
+        // ensure the container is not running, ignore if not present
+        DockerTestUtils.execute("docker", "kill", CONTAINER_NAME).getExitValue();
+    }
+
+    private static void checkpointTestProcess() throws Exception {
+        int javaPid = DockerTestUtils.execute("docker", "exec", CONTAINER_NAME, "/jdk/bin/jps").asLines().stream()
+                .filter(line -> !line.isBlank() && !line.contains("Jps"))
+                .mapToInt(line -> Integer.parseInt(line.split("\\s")[0]))
+                .min().orElseThrow();
```
Comment (AntonKozlov) [inline test/jdk/jdk/crac/TestInetAddress.java:130]: Not something really need to fix, but it's possible to specify just e.g. class-name and not parse jps output: https://github.com/openjdk/crac/blob/master/src/jdk.jcmd/share/man/jcmd.1#L58

File: src/java.base/share/classes/java/net/InetAddress.java [hunk 3]
```diff
@@ -341,6 +347,20 @@ public byte[] addressBytes(Inet6Address inet6Address) {
...
+            public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+                cache.clear();
+                expirySet.clear();
+            }
+
+            @Override
+            public void afterRestore(Context<? extends Resource> context) throws Exception {
+            }
+        };
+        Core.getGlobalContext().register(checkpointListener);
```
Comment (AntonKozlov) [inline src/java.base/share/classes/java/net/InetAddress.java:363]: For JDK internal Resources, please prefer jdk.internal.crac.Core.getJDKContext() https://github.com/openjdk/crac/blob/crac/src/java.base/share/classes/jdk/internal/crac/Core.java#L41

**Review by AntonKozlov — COMMENTED** [review]
Looks very nice! I don't see this to the Draft anymore :)

**Comment by rvansa** [comment]
@AntonKozlov Thanks for the review, updated.

**Comment by rvansa** [comment]
GHA failures should be fixed by https://github.com/openjdk/crac/pull/44

**Review by AntonKozlov — APPROVED** [review]
Thank you! LGTM

File: test/jdk/jdk/crac/java/net/InetAddress/ResolveTest.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.function.Consumer;
+
+/*
+ * @test
+ * @summary Test if InetAddress cache is flushed after checkpoint/restore
+ * @requires docker.support
+ * @library /test/lib
+ * @modules java.base/jdk.crac
+ * @build ResolveInetAddress
+ * @run main/timeout=360 TestInetAddress
```
Comment (wkia) [inline test/jdk/jdk/crac/java/net/InetAddress/ResolveTest.java:48]: I guess TestInetAddress needs to be changed here also. Currently the test fails due to absense of TestInetAddress

Comment (rvansa) [reply test/jdk/jdk/crac/java/net/InetAddress/ResolveTest.java:48]: Oops, thanks for spotting this. Looks like I still had the old class cached when running the test...

**Review by AntonKozlov — APPROVED** [review]
I've run the new test on the latest revision and it passes

**Pull request closed without merging**